src/uct/tcp/tcp_cm.c

```diff
@@ -8,100 +8,289 @@
-    ep->conn_state = new_conn_state;
+    status = uct_tcp_cm_recv_conn_ack(ep);
+    if (ucs_unlikely(status != UCS_OK)) {
+        uct_tcp_ep_mod_events(ep, 0, EPOLLIN);
+        return ret;
+    }
+
+    ret++;
+
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+    uct_tcp_ep_mod_events(ep, EPOLLOUT, EPOLLIN);
+
+    ucs_assertv(ep->tx.buf == NULL, "ep=%p", ep);
+
+    /* Progress possibly pending TX operations */
+    return ret + uct_tcp_ep_progress(ep, UCT_TCP_EP_CTX_TYPE_TX);
+}
 
-    if (!ucs_log_is_enabled(UCS_LOG_LEVEL_DEBUG)) {
-        return;
+unsigned uct_tcp_cm_conn_req_rx_progress(uct_tcp_ep_t *ep)
+{
+    unsigned ret = 0;
+    struct sockaddr peer_addr;
+    ucs_status_t status;
+
+    status = uct_tcp_cm_recv_conn_req(ep, &peer_addr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        goto err;
     }
 
-    ucs_sockaddr_str((const struct sockaddr*)&iface->config.ifaddr,
-                     str_local_addr, UCS_SOCKADDR_STRING_LEN);
-    ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr, UCS_SOCKADDR_STRING_LEN);
+    ret++;
 
-    ucs_debug("tcp_ep %p: %s -> %s for the [%s]<->[%s] connection",
-              ep, uct_tcp_ep_cm_state[old_conn_state].name,
-              uct_tcp_ep_cm_state[ep->conn_state].name,
-              str_local_addr, str_remote_addr);
+    /* Cleanup the address that was allocated during the connection
+     * accepting and assign the address received from the connection
+     * request message */
+    uct_tcp_ep_addr_cleanup(&ep->peer_addr);
+    status = uct_tcp_ep_addr_init(&ep->peer_addr, &peer_addr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        goto err;
+    }
+
+    status = uct_tcp_cm_send_conn_ack(ep);
+    if (ucs_unlikely(status != UCS_OK)) {
+        goto err;
+    }
+
+    ret++;
+
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+
+    return ret;
+
+ err:
+    uct_tcp_ep_mod_events(ep, 0, EPOLLIN);
+    uct_tcp_ep_destroy(&ep->super.super);
+    return ret;
 }
 
 ucs_status_t uct_tcp_cm_conn_start(uct_tcp_ep_t *ep)
 {
-    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
-                                            uct_tcp_iface_t);
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
     uct_tcp_ep_conn_state_t new_conn_state;
+    uint32_t req_events;
     ucs_status_t status;
 
     status = ucs_socket_connect(ep->fd, ep->peer_addr.addr);
-    if (status == UCS_INPROGRESS) {
-        iface->outstanding++;
-
+    if (ucs_likely(status == UCS_INPROGRESS)) {
         new_conn_state  = UCT_TCP_EP_CONN_STATE_CONNECTING;
+        req_events      = EPOLLOUT;
         status          = UCS_OK;
-
-        uct_tcp_ep_mod_events(ep, EPOLLOUT, 0);
     } else if (status == UCS_OK) {
-        new_conn_state  = UCT_TCP_EP_CONN_STATE_CONNECTED;
+        status = uct_tcp_cm_send_conn_req(ep);
+        if (status != UCS_OK) {
+            ucs_error("tcp_ep %p: failed to initiate the connection "
```

Comment:
need to print this error inside uct_tcp_cm_send_conn_req()
pls see https://github.com/openucx/ucx/blob/master/doc/CodeStyle.md#error-handling
also would reuse code in other call to uct_tcp_cm_send_conn_req()

Reply:
done